### ISSUE TITLE
New-UDIcon: Ability to specify custom icons

### ISSUE THREAD

**Issue by donlent** [ironmansoftware/powershell-universal#640]
### Summary of the new feature / enhancement

It would be great to specify a custom icon within New-UDIcon. For example by using *.ico Files that can be put into an asset folder or even by specifying a web link for the location.

Reason why this would be good:
In case you want to develop a dashboard / page with customer specific icons because they are related to applications / symbols that the users are used to, this would make the UI even more intuitive for the users.

### Proposed technical implementation details (optional)

New-UDIcon -Icon '$assetFolder\myIcon.ico'

New-UDIcon -Icon 'https://myWebRepo/myIcon.ico'

**Comment by adamdriscoll** [comment]
You can use `New-UDImage` to display icons. 

```
New-UDImage -Url "favicon.ico"
```

**Comment by donlent** [comment]
Hey @adamdriscoll 
Thanks for your answer.

But this can't be used within other elements right? Such as for a assigning a custom icon on a button or a listitem...?
In that case it wouldn't be what I am looking for... 😉

**Comment by rwidmark** [comment]
Yeah, this would be nice.
I do agree with @donlent that you should implement a searchpath for the icon in the New-UDIcon because then you can use it with other PSU components like buttons and add it to list etc as you can do today with new-udicon.

It would be great if it was possible to both use weburl but also local search path as it would be nice to store icons locally on the PSU server as well in a own folder or on a share.

It's also important that you still can use the other parameters that New-UDIcon offers with the "custom icons" for example -Size , -Style , -Color etc.

Do you agree @adamdriscoll or do you have something else in mind?

**Comment by Omzig** [comment]
@adamdriscoll, it would be nice to be able to give it a fontawesome key of some kind and then that opens up Pro icons, and if you have a ProMax you can upload your own to their service.

**Comment by donlent** [comment]
Dear @adamdriscoll 
Thanks a lot for having included that into Milestone 4.0! Finally! 🥳

I understand that you'll have to remove FontAwesome Pro Icons from your native integration because that probably violates the license scope.
However I also do agree that beside the possibility to specify custom icons we should also have the possibility to buy a FA Pro License and integrate it with PSU. Like that they get their money for each PSU customer that want's to use it and you're safe with your product. :-)

So to wrap up this feature requests consists of two topics:

- [ ] Allow us to define custom icons (*.ico is probably the most easy way to go. since you can embed various sizes. *.png may also work but in my opintion it's not a priority)
- [ ] Pay attention that the functionality of New-UDIcon is as transparent as possible with custom icons ( for example -Size)
- [ ] Allow us to specify a FontAwesome Pro Key to use the Icons Pro Set and variants

If I forgot something feel free to add your comments.

Thanks again,
Don

**Comment by adamdriscoll** [comment]
After looking into a lot of the popular icon sets available out there, we've decided to make this super generic. With each icon set being totally different in how you use them, we need a way to be able to allow users to just customize whatever they want in an icons. The implementation just allows specifying content for the icons. 

Here's an example of using the css.gg icon set. 

```
        New-UDHelmet -Tag 'link' -Attributes @{
            href = 'https://css.gg/css' 
            rel = 'stylesheet'
        }

        New-UDIcon -Children {
            New-UDHtml '<i class="gg-adidas"></i>'
        }
```

![image](https://user-images.githubusercontent.com/1256531/233131939-26068548-cbcd-498f-86b9-e03130adc8f9.png)

This implementation will also allow you to use ico files loaded as images or other icon fonts. We've also included UDHelmet directly in PSU to make it easier to load external files. 

Due to licensing issues with FontAwesome Pro, we cannot ship the FontAwesome Pro CSS and JS files with PSU; even without a pro key. That said, it's possible to purchase FontAwesome Pro and then use the exact same method we used above for css.gg to load the external css\js files from the FA CDN and then display the pro icons.

**Comment by rwidmark** [comment]
@adamdriscoll what kind of sites do you recommend for icons?

Also, how can you change color and size on the icon?
I'm guessing that's the thing people will miss

**Comment by adamdriscoll** [comment]
These are the ones I looked at. 

- https://feathericons.com/
- https://css.gg/
- https://ionic.io/ionicons/
- https://icons.getbootstrap.com/
- https://github.com/vorillaz/devicons

Each library seems to have it's own method of styling. 

http://vorillaz.github.io/devicons/#/usage
https://icons.getbootstrap.com/#styling
https://ionic.io/ionicons/usage
https://boxicons.com/usage

react-icons has a great list of integrations that also serves as a good list of icon libraries. 

https://react-icons.github.io/react-icons/icons?name=bi

**Comment by rwidmark** [comment]
@adamdriscoll can you provide a quick example how to set color and size when using your example above?

Also, is it possible to use this for a local stored icon set today with 3.8.10 or is it something that will be possible in 4.0?

As I understands it now I need to install the udhelmet module and then import it in my dashboard to make this possible today with 3.8.10 (the example above)

**Comment by rwidmark** [comment]
@adamdriscoll  Your code above won't work as it says that -Tag are missing and as far as I can see New-UDHelmet don't contain this either.

**Comment by adamdriscoll** [comment]
This requires v4.

**Comment by rwidmark** [comment]
> This requires v4.

Ok, I see :) what beta version? I'm thinking that I might need to start as soon as possible as the Pro icons are going to be removed.

**Comment by donlent** [comment]
Hi @adamdriscoll 
Thanks a lot for your update on this one.
Sounds interesting but I'm not quite sure to understand right now how to use this technique to use various icon sets in an App (Dashboard ;-) )? I think to understand that with UD-Helmet you basically override the link tag of the Body... so how can we define multiple UD-Helmets for various icon sets? Will that be possible?

And also:
How to use locally stored icons? 

Basically all my questions relate to one important thing:
Please make sure that this new feature will be well documented in order to avoid confusion when using fontawesome UD-Icons vs. custom ones. Also showing multiple use case scenarios will be important.

Thanks a lot for that in advance!

**Comment by adamdriscoll** [comment]
@rstolpe - This is in beta2 and will be in the current nightly build.
@donlent - I agree that showing multiple use cases will be important. You will be able to use multiple helmets for multiple tags if you want. Storing icons locally will likely require a published folder with your icon sets in them. I'll make sure we capture all that in the docs when we release this.

**Comment by donlent** [comment]
Thank you very much @adamdriscoll 🙏🏻🤗
I'n sure this will be a big help for everybody...

**Comment by rwidmark** [comment]
@adamdriscoll is it possible to implement this in 3.8.11 as pro icons will be removed in that version?
I have created a new issue for that as it's a issue when a feature gets removed without a replacement feature

**Comment by rwidmark** [comment]
@adamdriscoll any news or similar for the udhelmet thing?
Do you have any article or other explanation?
